### PR TITLE
[BEAM-6122] Update committer guidelines

### PR DESCRIPTION
**Pull request by tweise**
Update guidelines per discussion in https://lists.apache.org/thread.html/6d922820d6fc352479f88e5c8737f2c8893ddb706a1e578b50d28948@%3Cdev.beam.apache.org%3E

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: website/src/contribute/committer-guide.md [hunk 1]
```diff
@@ -24,6 +24,24 @@ This guide is for
...
+  but committer may absorb some extra effort for new contributors)
+* Pull requests and commit messages establish a clear history with purpose and origin of changes
+* Ability to perform a granular rollback, if necessary (also see [policies](/contribute/postcommits-policies/))
+
+Granularity of changes:
+
+* We prefer small independent, incremental PRs with descriptive, isolated commits. Each commit is a single clear change
+* It is OK to keep separate commits for different logical pieces of the code, if they make reviewing and revisiting code easier
+* Making commits isolated is a good practice, authors should be able to relatively easily split the PR upon reviewer's request
+* When there are multiple commits in a single PR, every commit that gets merged should compile and pass tests
```
Comment (robertwb) [inline website/src/contribute/committer-guide.md]: This is too strong, sometimes it is really helpful to split out completely automated (but possibly broken) refactorings from the manual fixups that were required.

Comment (tweise) [reply website/src/contribute/committer-guide.md]: Do you have a suggestion how to word it? The challenge would be to come up with something that is clear and actionable for the common case.

File: website/src/contribute/committer-guide.md [hunk 2]
```diff
@@ -77,15 +95,25 @@ At some point in the review process, the change to the codebase will be
...
+The committer may use the "Squash and merge" option in Github (or modify the PR commits in other ways).
+The committer is ultimately responsible and we "trust the committer's judgment"!
 
 ## Merging it!
 
 After all the tests pass, there should be a green merge button at the bottom of
-the pull request.  There are multiple choices and you should choose "Merge pull
+the pull request. There are multiple choices. Unless you want to squash commits
+as part of the merge (see above) you should choose "Merge pull
 request" (the default). This preserves the commit history and adds a merge
```
Comment (apilloud) [inline website/src/contribute/committer-guide.md]: `Merge pull request` actually defaults to whatever operation you did last time. Can you remove `(the default)` and replace it with `and ensure "Create a merge commit" is selected from the drop down.`

File: website/src/contribute/committer-guide.md [hunk 3]
```diff
@@ -100,3 +128,6 @@ Instead, pull it all into the subject line:
     Merge pull request #1234: [BEAM-7873] Fix the foo bizzle bazzle
 
 If you have comments to add, put them in the body of the commit message.
+
+After merging the pull request, the committer is responsible for resolving associated JIRA(s)
```
Comment (swegner) [inline website/src/contribute/committer-guide.md]: Do you mean the contributor's responsibility, or the JIRA assignee? The contributor has the most context on whether the PR completes the associated JIRA. It might be that a JIRA requires multiple PR's (or something external, like a release) to complete.

I think it would be difficult for the committer to take this responsibility.

File: website/src/contribute/committer-guide.md (hunk 1 above)
Comment (kennknowles) [reply website/src/contribute/committer-guide.md]: I actually experimented with git aliases / scripts to do this automatically. It isn't terribly hard for precommits - you iterate over all unmerged commits and force push them one by one. This might flood Jenkins a bit. It would be nice to have this for postcommits. It seems like it ought to be a checkbox in Jenkins config actually.

I also want to mention that the rollback PR should be green except in rare cases. So a green rollback of one commit is just great.

Comment (kennknowles) [reply website/src/contribute/committer-guide.md]: I would simply remove this bullet. I don't think there's anything that is both reasonable and actionable.

In practice, one may often revert a whole PR as Scott suggests. You can always open multiple reverts for the whole PR, various commits in the PR, etc, and choose one (the first?) that restores signal. It is then easy enough to roll forward innocent commits that were caught in the dragnet.

File: website/src/contribute/committer-guide.md [hunk 4]
```diff
@@ -77,15 +95,25 @@ At some point in the review process, the change to the codebase will be
...
 commits that are not meaningful to preserve in the history. The reviewer should
 give the LGTM and then request that the author of the pull request rebase,
-squash, split, etc, the commits, so that the history is most useful. Favor
-commits that do just one thing. The commit is the smallest unit of easy
+squash, split, etc, the commits, so that the history is most useful:
+* Favor commits that do just one thing. The commit is the smallest unit of easy
 rollback; it is easy to roll back many commits, or a whole pull request, but
 harder to roll back part of a commit.
+* Commit messages should tag JIRAs and be otherwise descriptive.
+It should later not be necessary to find a merge or first PR commit to find out what caused a change.
```
[the remark this answers is not in the document]
Comment (kennknowles) [reply website/src/contribute/committer-guide.md]: Yea, GitHub has really stepped up this UI support. That said, `git log` is much more powerful in a lot of ways so it should be first class too.

File: website/src/contribute/committer-guide.md [hunk 5]
```diff
@@ -77,15 +95,25 @@ At some point in the review process, the change to the codebase will be
...
+* Squash the "Fixup!", "Address comments" type of commits that resulted from review iterations.
+
+While it is preferred that authors squash commits after review is complete,
```
Comment (kennknowles) [inline website/src/contribute/committer-guide.md]: Should this be in "Merging it!" section below?

File: website/src/contribute/committer-guide.md (hunk 3 above)
Comment (kennknowles) [reply website/src/contribute/committer-guide.md]: Yea, I think it should be the contributor who did the change and presumably has the Jira assigned to them. We do have a problem with Jira's staying open until they get re-triaged but not sure that fits with the discussion we had on list.

File: website/src/contribute/index.md [hunk 6]
```diff
@@ -149,9 +149,11 @@ To contribute code, you need
    have an open source license [compatible](https://www.apache.org/legal/resolved.html#criteria) with Apache.
 1. Add unit tests for your change
 1. When your change is ready to be reviewed and merged, create a pull request.
-   Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`,
+   Format commit messages and the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`,
```
Comment (kennknowles) [inline website/src/contribute/index.md]: I think some time ago we had a line that was something like "for trivial changes this is not necessary", when the rule applied only to pull requests. Is that line deleted? Just checking. I am actually fine having it deleted. I think that someone should be very involved in the project enough to have appropriate judgment about when it might be overkill.

File: website/src/contribute/index.md [hunk 7]
```diff
@@ -172,6 +174,9 @@ To contribute code, you need
    Use `R: @username` in the pull request to notify a reviewer.
 1. If you don't get any response in 3 business days, email the dev@ list to ask for someone to look at your pull
    request.
+1. Review feedback typically leads to follow-up changes. Please add these changes as additional "fixup" commits to the
+   existing PR/branch. This will allow reviewer(s) to track the incremental progress. After review is complete and the
+   PR accepted, multiple commits should be squashed (see [Git workflow tips](https://cwiki.apache.org/confluence/display/BEAM/Git+Tips)).
```
Comment (kennknowles) [inline website/src/contribute/index.md]: The number of links into the Wiki is getting worrisome, as it is unstable by design. I wonder if there is another way to handle this?

File: website/src/contribute/index.md [hunk 8]
```diff
@@ -172,6 +174,9 @@ To contribute code, you need
    Use `R: @username` in the pull request to notify a reviewer.
 1. If you don't get any response in 3 business days, email the dev@ list to ask for someone to look at your pull
    request.
+1. Review feedback typically leads to follow-up changes. Please add these changes as additional "fixup" commits to the
```
Comment (kennknowles) [inline website/src/contribute/index.md]: I don't think we agreed on this, did we? This used to be critical for GitHub to not lose comments, but I think that is no longer the case. I see a lot of PRs that don't do this and it seems OK. In fact, when a large PR is reviewed incrementally like this I think it tends to let in worse changes because the increments after the first round look OK even though the resulting code is not so good.

FWIW reviewable.io does not require this and has its own incremental review that we have toyed with. Anyone can choose to use it on a per-review basis.

File: website/src/contribute/committer-guide.md (hunk 3 above)
Comment (amaliujia) [reply website/src/contribute/committer-guide.md]: It's a trade off between full context of a JIRA and a well managed JIRA task list. Contributors could have more context on a JIRA but contributors could leave the JIRA open unintentionally. 

A compromise could be that committer asks in a JIRA if it can be close after approve the PR, and close the JIRA if no reply shows up within a reasonable time (like 3 days). Contributor could always reopen the JIRA or add more context later.

File: website/src/contribute/committer-guide.md (hunk 1 above)
Comment (charlesccychen) [reply website/src/contribute/committer-guide.md]: I would like to +1 removing this point.  It's not feasible to trigger precommits for each commit, and if we find that tests don't pass for the overall PR, it is too much of a hassle to suggest that the author should rewrite the history to satisfy this point (and in practice I think this will be ignored).

I would also discourage reverting part of a PR and suggest that for simplicity, the entire PR should be the unit of reversion (say if a PR is found to break tests); a subsequent PR can roll forward partial changes if necessary.

File: website/src/contribute/committer-guide.md (hunk 3 above)
Comment (amaliujia) [reply website/src/contribute/committer-guide.md]: And yes, seems like this point is not on the discussion thread. We might need to have a discussion for it since it affects both roles (contributor and committer).

File: website/src/contribute/committer-guide.md (hunk 1 above)
Comment (kennknowles) [reply website/src/contribute/committer-guide.md]: I'd be cautious about the phrasing of your last paragraph. We should not force submit reverts. The revert itself needs to be green, with extraordinarily rare exceptions. Otherwise it has not achieved its primary purpose, which is to restore signal. So there's no added simplicity nor time saved by making rules about what you should revert. The important distinction of a rollback is that you don't wait for a rollforward.

**Comment by swegner** [comment]
In general, I believe that manual processes should be made as simple as possible and have failure-modes which give timely and actionable feedback.

Some of these policies (like tagging a JIRA on every commit) would be very easy to validate in a pre-commit ([the website build.gradle gives an example of interacting with Git from Gradle](https://github.com/apache/beam/blob/3a4da9cd9459605c0086c2db57eae680e0862d43/website/build.gradle#L194)). But, the workflow for fixing up commit messages takes some degree of Git mastery (`git rebase -i`). It's not a huge leap, but it adds the requirement that all contributors known intermediate/advanced Git (or always write compliant commit messages on the first try).

A simpler failure mode would be something like: "*If each commit is tagged with a JIRA, individual commits will be retained during merge; otherwise commiters should squash and merge a single commit with JIRA information from the PR.*"

**Comment by kennknowles** [comment]
I really like Scott's latest points. Another aspect here is that the instructions for new contributors should be primarily helpful and welcoming, and avoid being prescriptive or pedantic. And _concise_. We want them to have an easy onramp, so we need enough tips, but we also don't want to discourage their contributor either by excessive demands or an intimidatingly large amount of instruction. My biggest goal is for them to open a PR and have a good interaction with a reviewer. I don't want to put anything in the way of that. That's why rules for policing our more experienced members should be clearly separate. As an extremely common special case, if someone thinks they already know how to open a PR, I want to get out of the way and let them do that, and work out the rest once we are in conversation. Putting this stuff in Jenkins is great not only for the usual reasons for automation but also because it makes iteration on it self-service.

Knowledgeable committers can take on additional chores and/or chat with the contributor about things on the PR. I do think Beam committers can be expected to know `rebase -i` and how to push to a contributor's branch, or else abstain from merging PRs. 

On the automation front, just noting that we have 1575 non-merge commits on `master` since Oct 1 (choosing a recent date to avoid any differences in older policies) of which 454 are tagged with a Jira. I don't think that 2/3 of the commits are out of line. But I also think forcing folks to just tag them with the Jira they happen to also be addressing is fine. It is a really nice way to handle the "should I [ask for] squash or not?" problem.

File: website/src/contribute/index.md (hunk 8 above)
Comment (tweise) [reply website/src/contribute/index.md]: I don't think it was part of the discussion, rather it seems to be the undocumented status quo from looking at what happens currently (changes being added incrementally to PRs)? I don't have a strong preference one way or the other.

File: website/src/contribute/index.md (hunk 6 above)
Comment (tweise) [reply website/src/contribute/index.md]: Not aware of the history of that line. But in the spirit of keeping instructions concise, we should probably leave the expert mode out from here?

File: website/src/contribute/committer-guide.md (hunk 3 above)
Comment (tweise) [reply website/src/contribute/committer-guide.md]: We should move this line out of the PR and discuss it separately. To review a PR, I probably need to sufficiently understand the scope of the JIRA and what the change I just merge addresses or not. Typically the committer merging the PR takes the last action in the chain and the contributor may not return just to close out the JIRA. The committer is also more likely to be aware of the versioning.

File: website/src/contribute/committer-guide.md (hunk 1 above)
Comment (tweise) [reply website/src/contribute/committer-guide.md]: Looks like there is agreement to just drop this item.

Comment (mxm) [reply website/src/contribute/committer-guide.md]: The beauty of every PR compiling and passing tests is that you can use `git-bisect` without any trouble. There might be a way to bisect only with merge commits; individual broken commits wouldn't matter then. 

Instead of dropping the paragraph we could rephrase it to "Ideally, every commit should compile and pass tests" but I generally agree that it is not practical nor do we have the right test infrastructure to verify it.

File: website/src/contribute/index.md (hunk 8 above)
Comment (robertwb) [reply website/src/contribute/index.md]: IIRC, github comments are tied to diffs, so they can be lost if the code changes too much and the commits are gone. I personally like seeing just the changes, as it gives me the choice of viewing everything vs. the incremental diff (and the default is still everything). If consensus isn't obvious here, maybe back to the list?

File: website/src/contribute/committer-guide.md (hunk 3 above)
Comment (robertwb) [reply website/src/contribute/committer-guide.md]: I'm also in favor of removing this line. I think it's nice to do (especially if it's obvious and you have the context in your head) but not essential.

File: website/src/contribute/index.md (hunk 7 above)
Comment (robertwb) [reply website/src/contribute/index.md]: For this content, I'd rather have links that might go bad than not have them (or a vague check the wiki).

File: website/src/contribute/committer-guide.md (hunk 1 above)
Comment (tweise) [reply website/src/contribute/committer-guide.md]: Changed wording as suggested by Robert.

File: website/src/contribute/committer-guide.md (hunk 2 above)
Comment (tweise) [reply website/src/contribute/committer-guide.md]: Done!

File: website/src/contribute/committer-guide.md (hunk 3 above)
Comment (tweise) [reply website/src/contribute/committer-guide.md]: Removed.

File: website/src/contribute/index.md (hunk 8 above)
Comment (tweise) [reply website/src/contribute/index.md]: Changed wording.

File: website/src/contribute/committer-guide.md (hunk 1 above)
Comment (tweise) [reply website/src/contribute/committer-guide.md]: > I also want to mention that the rollback PR should be green except in rare cases.

Added that under rollback.

File: website/src/contribute/committer-guide.md (hunk 5 above)
Comment (tweise) [reply website/src/contribute/committer-guide.md]: Done!

**Comment by tweise** [comment]
@swegner @kennknowles PTAL
